Chatkit/Subscriptions/Subscription.swift

```diff
@@ -0,0 +1,301 @@
+            state = .subscribingStageOne(instanceWrapper: instanceWrapper,
+                                         completions: completions.appending(completion))
+            
+        case let .subscribingStageTwo(instanceWrapper, resumableSubscription, completions):
+            // Add the new completion handler to the `completions` of the .subscribing state
+            state = .subscribingStageTwo(instanceWrapper: instanceWrapper,
+                                         resumableSubscription: resumableSubscription,
+                                         completions: completions.appending(completion))
+            
+        case .subscribed:
+            // Immediately invoke success, we're already subcribed
+            completion(.success)
+            
+        }
+    }
+    
+    func unsubscribe() {
+        
+        switch state {
+            
+        case .notSubscribed:
+            // Nothing to do
+            ()
+            
+        case let .subscribingStageOne(_, completions):
+            
+            state = .notSubscribed
+            
+            let error = Self.unsubscribeCalledWhileSubscribingError
+            
+            delegate?.subscription(self, didReceiveError: error)
+            
+            for completion in completions {
+                completion(.failure(error))
+            }
+            
+        case let .subscribingStageTwo(_, resumableSubscription, completions):
+            
+            resumableSubscription.onOpen = nil
+            resumableSubscription.onError = nil
+            resumableSubscription.end()
+            
+            state = .notSubscribed
+            
+            let error = Self.unsubscribeCalledWhileSubscribingError
+            
+            delegate?.subscription(self, didReceiveError: error)
+            
+            for completion in completions {
+                completion(.failure(error))
+            }
+            
+        case let .subscribed(_, resumableSubscription):
+            
+            resumableSubscription.onOpen = nil
+            resumableSubscription.onError = nil
+            resumableSubscription.end()
+            
+            state = .notSubscribed
+        }
+    }
+    
+    // MARK: - Private
+    
+    private func performSubscribe(completion: @escaping SubscribeHandler) {
+        
+        let instanceWrapper = dependencies.instanceWrapperFactory.makeInstanceWrapper(forType: .subscription(subscriptionType))
+        
+        let requestPath = makeRequestPath(for: subscriptionType)
+        let requestOptions = PPRequestOptions(method: HTTPMethod.SUBSCRIBE.rawValue, path: requestPath)
+        
+        let onEvent: InstanceWrapper.OnEvent = { [weak self] _, _, jsonDataAsAny in
+            guard let self = self else {
+                return
+            }
+            
+            guard let jsonData = jsonDataAsAny as? Data else {
+                preconditionFailure("`onEvent` called with non-`Data` typed data")
```

Comment:
`PusherChatkit` is a third party SDK, therefore we should aim to always fail gracefully. In this case we should probably throw an error instead of crashing the whole app.

Reply:
The same applies other uses `preconditionFailure` and similar methods outside our testing targets.

Reply:
Sorry agreed this is pretty ugly.  I've changed all `preconditionFailure`s to `assertionFailure`.

Regarding the guard, this is only necessary because of PusherPlatform's failings which should get addressed later as part of "[PusherPlaform Updates](https://pusherpm.leankit.com/card/1002830214)" so I've added a comment to suggest this.